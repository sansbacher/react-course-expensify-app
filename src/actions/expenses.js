import database from '../firebase/firebase'

// NEW, with redux-thunk, for use with Firebase or other async APIs:
// Async Actions workflow, eg. for use with DBs:
// 1. Component calls Action generator
// 2. Action generator returns a Function
// 3. Component dispatches the function
// 4. Function runs (has the ability to dispatch other actions or do whatever it wants)

// OLD, plain old local redux store - see _SAVE version for this workflow:
// Basic Actions workflow:
// 1. Component calls Action generator
// 2. Action generator returns an Object (eg. from this file, or the filters actions)
// 3. Component dispatches the object
// 4. The Redux store changes

export const addExpense = expense => ({				// Simplified addExpense Action Generator
	type: 'ADD_EXPENSE',
	expense
})

export const startAddExpense = (expenseData = {}) => {			// Returns a Function (which requires redux-thunk to process)
	return (dispatch) => {		// Redux-Thunk will call this function with dispatch
		const {					// Destructure, providing defaults as needed
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData
		const expense = {description, note, amount, createdAt}		// Construct new Object (includes the defaults)

		// Pushes the data into Firebase DB - which is async - then dispatch the normal addExpense Action. Returns a promise, which can be chained upon.
		return database.ref('expenses').push(expense).then((ref) => {
			dispatch(addExpense({
				id: ref.key,			// ID is no longer a uuid v4, but is now the unique 'key' from Firebase
				...expense				// All the other items from expense
			}))
		});
	}
}

export const removeExpense = ({id} = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
})

export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
})

