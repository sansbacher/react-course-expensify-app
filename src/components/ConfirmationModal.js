import React from 'react'
import Modal from 'react-modal'

// Tell React-Modal what element (our root div) to hide when displaying the modal for assistive readers:
// http://reactcommunity.org/react-modal/accessibility/
if (process.env.NODE_ENV !== 'test') {						// To avoid errors when testing with Jest: https://github.com/reactjs/react-modal/issues/632
	Modal.setAppElement('#app')
}

const ConfirmationModal = props => (
	<Modal
		isOpen={props.showConfirmation}
		contentLabel="Confirm Remove"
		closeTimeoutMS={150}
		className="modal"
	>
		<h3 className="modal__title">Remove Expense?</h3>
		<p className="modal__body">Are you sure you want to remove this Expense:</p>
		<p className="modal__body--message">{props.expenseDescription}</p>
		<div>
			<button className="button" onClick={props.handleRemove}>Remove</button>
			<button className="button button--secondary" onClick={props.handleCancel}>Cancel</button>
		</div>
	</Modal>
)

export default ConfirmationModal;