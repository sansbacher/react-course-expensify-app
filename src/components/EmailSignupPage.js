import React from 'react'
import { Link } from 'react-router-dom';

export class EmailSignupPage extends React.Component {

	render() {
		return (
			<div className="box-layout">
				<div className="box-layout__box box-layout__box--large">
					<h1 className="box-layout__title">Signup with your Email Address</h1>
					<p>Enter your email address</p>	
					<p>Enter your password</p>
					<p>etc...</p>
					<br />
					<p>Have an account? <Link to="/">Log in</Link></p>
				</div>
			</div>
		);
	}
}

export default EmailSignupPage;