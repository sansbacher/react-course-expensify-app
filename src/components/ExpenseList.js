import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// Regular unconnected Component, once connected it will receive the mapped State to Props
// ONLY exported for the purposes of TESTING with Jest (NOT used directly by the app, the default Connected export below is)
export const ExpenseList = props => (
	<div>
		{/* In component: {...expense} JSX spreads properties to individual props, like amount={expense.amount} */}
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			) : (
				props.expenses.map(expense => (
					<ExpenseListItem {...expense} key={expense.id} />
				))
			)
		}
	</div>
);

// Also receives props, so could use: (state, props)
const mapStateToProps = (state) => {
	// Returns an object of what parts of state (that will be added to props) that will be used by the function connect() returns
	return {															// .dispatch() is also added to the returned props
		expenses: selectExpenses(state.expenses, state.filters)			// Handles filtering/sorting of the state
	};
};

// connect() returns a function that we're immediately executing, Like a IIFE.
// We pass to connect() a function that returns what part of the passed in state that gets used in the returned 'IIFE' function (added to props)
// The 'IIFE' returns a Higher Order Component [HOC] (which is a wrapped version of our component: the Connected Component)
// The default export is the function returned by the call to the function returned by connect()
// That function can access whatever we select from the store, and will react as the store changes
/* 
	// This is the verbose version:
	let makeHOCWithState = connect(mapStateToProps)
	let WrappedConnectedComponent = makeHOCWithState(ExpenseList)
	export default WrappedConnectedComponent;
*/
// Technically the mapStateToProps function isn't required, connect()(SomeComponent) is valid - just means only .dispatch() is added to props
// At it's basic level: connect() is a method to inject/add part of state (and dispatch) into the props for another function
export default connect(mapStateToProps)(ExpenseList);
