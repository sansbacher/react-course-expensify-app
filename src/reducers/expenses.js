
const expensesReducerDefaultState = []

// Was: expensesReducer()
export default (state = expensesReducerDefaultState, action) => {			// For this reducer state is JUST the array (not the whole combinedState)
	switch (action.type) {
		case 'ADD_EXPENSE':
			// return state.concat(action.expense);			// state.push() would change the state. concat() returns a new array.
			return [...state, action.expense];				// Or we can use spread operator, returning a new array with the existing state + new item
		case 'REMOVE_EXPENSE': 
			return state.filter(expense => expense.id !== action.id)
		case 'EDIT_EXPENSE':
			return state.map((expense) => {
				if (expense.id === action.id) {
					return {
						...expense,							// spread existing expense
						...action.updates					// plus any updates
					};
				} else {
					return expense;
				}
			})
		default:
			return state;
	}
}
