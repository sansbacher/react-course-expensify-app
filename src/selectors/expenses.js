import moment from 'moment'

// Was: getVisibleExpenses()
export default (expenses, {text, sortBy, startDate, endDate}) => {
	// Filter by Text acts on Description
	// Filter by Date acts on createdAt
	// Sort by date = most recent expense first
	// Sort by amount = largest expense first
	return expenses.filter((expense) => {
		const createdAtMoment = moment(expense.createdAt)
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true							// If no startDate then true, to not filter on it
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
		const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

		return startDateMatch && endDateMatch && textMatch;
	}).sort((a, b) => {											// Compare function: -1 if a before b, 1 if a after b, 0 if a == b (same)
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;			// Most recent expense first
		} else if (sortBy === 'amount') {
			return a.amount < b.amount ? 1 : -1;				// Largest expense first
		}
	});
} 