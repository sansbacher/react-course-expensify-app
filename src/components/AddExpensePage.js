import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {addExpense} from '../actions/expenses'

// Named export so we can test it
export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {		// Callback for when the form is Submitted (called by ExpenseForm )
		// Do these things when the form is is submitted:
		// this.props.dispatch(addExpense(expense))		// (OLD WAY!) Expects an Object, call dispatch() with the expense
		this.props.addExpense(expense)						// uses mapDispatchToProps (more testable as addExpense is encapsulated below) - call the function
		this.props.history.push('/')						// Redirect to the Dashboard (via React Route). Preserves history (Back works)
	}
	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		)
	}
}

// Similar to mapStateToProps we get access to pass in custom props, abstracting dispatch() away from the Component itself
const mapDispatchToProps = (dispatch) => {
	// Return an object of new props to add
	return {
		addExpense: expense => dispatch(addExpense(expense))		// addExpense: is a new prop that is a function calling dispatch() with addExpense baked in
	}
}

// Without any mapStateToProps function passed to connect() (ie. undefined or just nothing) it will only add .dispatch() to the props passed to ExpenseListItem
// BUT we can add mapDispatchToProps to customize props
export default connect(undefined, mapDispatchToProps)(AddExpensePage);