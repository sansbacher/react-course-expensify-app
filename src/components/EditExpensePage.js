import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense, startRemoveExpense} from '../actions/expenses'

// Named export just for testing
export class EditExpensePage extends React.Component {
	// Callbacks to handle what ExpenseForm or this form needs to do
	onSubmit = (expense) => {
		// OLD WAY (when it was a stateless functional component...): props.dispatch(editExpense(props.expense.id, expense))
		// OLD before Firebase: this.props.editExpense(this.props.expense.id, expense)			// Must use this.props.blah instead a class-based Component (not just props.blah)
		this.props.startEditExpense(this.props.expense.id, expense)			// Must use this.props.blah instead a class-based Component (not just props.blah)
		this.props.history.push('/')
	}
	onRemove = () => {
		// OLD WAY (...using embedded functions): this.props.dispatch(removeExpense({id: props.expense.id}))		// removeExpense() expects an {object} but only needs the id property
		// OLD before Firebase: this.props.removeExpense({id: this.props.expense.id})		// removeExpense() expects an {object} but only needs the id property
		this.props.startRemoveExpense({id: this.props.expense.id})		// startRemoveExpense() expects an {object} but only needs the id property
		this.props.history.push('/')
	}
	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm
						expense={this.props.expense}
						onSubmit={this.onSubmit}
						history={this.props.history}
					/>
					<button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
				</div>
			</div>
		)
	}
}

// the connect() function can have 0, 1, or 2 args (maybe others): none, state, or state + props
const mapStateToProps = (state, props) => {
	return {
		// Becomes props.expense or this.props.expense
		expense: state.expenses.find(expense => expense.id === props.match.params.id)
	};
}

// Encapsulate the call to dispatch() within a prop so we can more easily test this Component (could pass in: (dispatch, props) if needed)
const mapDispatchToProps = (dispatch) => {
	return {
		startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),		// Add these Props
		// startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))		// Destructure and restructure...
		startRemoveExpense: data => dispatch(startRemoveExpense(data))			// ... or just pass along the object passed in
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);