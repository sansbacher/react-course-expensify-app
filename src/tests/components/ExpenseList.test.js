// Jest provides test() as a global. BUT can use this to help VSCode out:
import {expect, test} from '@jest/globals'				// With React/Browser

import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'			// Need the UN-Connected 'raw' Component for Testing (we'll pass in dummy prop data)
import expenses from '../fixtures/expenses'							// Common dummy data

test('should render ExpenseList with expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={expenses} />)
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseList with empty message', () => {
	const wrapper = shallow(<ExpenseList expenses={[]} />)
	expect(wrapper).toMatchSnapshot()
})