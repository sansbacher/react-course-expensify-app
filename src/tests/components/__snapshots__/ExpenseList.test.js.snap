// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="gum"
    id={0}
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="credit card"
    id={2}
    key="2"
    note=""
  />
</div>
`;
